### ISSUE TITLE
I cannot access an opened sas session without recreating the object

### ISSUE THREAD

**Issue by Debbby57** [sassoftware/saspy#315]
Hello,
I have multiple python programs wich call SAS requests with saspy.
for each python program I have to open a new sas session. It's difficult to maintain the coherence with sas and python and I have to save my intermediary tables (in python or in sas) instead of leaving them in the work.

I would like to have a python program which open the sas session and the other python programs would retrieve this current session. Is it possible ? 

I imagine something like the getorcreate() method for sparksession :
program1
```python
def myfunction(kwargs**):
   sas_session=saspy.GetOrCreateSASSession()#if doesn't exist it's like the actual SASSession method 
   sas_session.submit("my code and save some tables in work")
```
program2
```python
def myotherfunction(kwargs**):
  sas_session=saspy.GetOrCreateSASSession() #here it gets the session opened by program1
  sas_session.submit("my code which could use  some tables in work")
```

**Comment by tomweber-sas** [comment]
Hey @Debbby57 , I can help you with this. So, are these different programs, like batch python scripts you run independently at different times? Or can you give me a better idea of the scope of the programs? This seems like one of these where a web-ex would be very useful as opposed to typing back and forth trying to figure this out. If you'd like to jump on a call and figure this out, I can set that up.
Tom

**Comment by Debbby57** [comment]
Hello @tomweber-sas, thanks for your reply !
I have a DAG with pipelines and tasks. I want the same SASSession for all my tasks in a pipeline.
We can do a webex for more details if you have time for that.

thank you for your help.

**Comment by tomweber-sas** [comment]
Cool. Sorry, was out yesterday on a holiday. How about this afternoon? 
Thanks,
Tom

**Comment by tomweber-sas** [comment]
So, just for future reference; putting this back on the thread (from email) so we have it here. We did a webex and see that Kedro is being used for building and running the pipeline. It didn't seem to have any way to allow for a common object to be referenced by the nodes in the pipeline, and adding the SASssion object to the inputs and outputs, to pass it along, was undesirable (understandably). So, afterward, I thought of this, which I think will get us what we were trying to do with their infrastructure, but couldn't. This, I believe, does what we were trying to do, but from the code we have control over.

from email:
I have an idea. This is simply just a scoping problem, and I can’t see anything in the Kedro doc
where we can solve this; there’s no scope nor way to access the SASsession object. So, what if
we just do that where you have control; your node functions. I’m thinking that instead of your node
functions just being stand alone function definitions, you define a class, with each of your node
functions being a method of the class. These methods, are the same code as the functions are currently;
same signatures and everything. But, because these are now methods of a class, in your __init__ of this class
you can assign 
self.sas = saspy.SASsession()
and then in your methods, you can access that as self.sas so that you can share that session.

With that class and the methods, you can now build your Pipeline. In the create pipeline, like we sort of tried,
you first instantiate your class (we were trying getting the SASsession here but had no way to reference it in the
node functions). Now that there’s an object of your class (let’s call it SASnodes) you can pass the methods to the
Node parameters in the pipeline definition, and then when they are run, they should be able to access self.sas.

def CreatePipeline(whatever):
     node_object = SASNodes()  # instantiate the object that has the SASsession, then your methods have access to it

     pipeline_obj = Pipeline(
                                      node(node_object.node1,   # since these are methods, when called they should have context to ‘self’
                                                 inputs=dict(a=1, b=1),
                                                 outputs=dict(cars_df='clean_cars2017',
                                                )
                                     node(node_object.node2, ,   # since these are methods, when called they should have context to ‘self’
                                                 inputs=dict(a=1, b=1),
                                                 outputs=dict(cars_df='clean_cars2017',
                                                )
                                      )

class SASNodes():
   def __init__(self, **kwargs):
      self.sas = saspy.SASsession()

   def node1(data: dict, **kwargs):
        df = self.sas.sd2df() # or whatever you need to do w/ SASsession
        return df

   def node2(data: dict, **kwargs):
        df = self.sas.sd2df() # or whatever you need to do w/ SASsession
        return df

I think this can work with the version you have now. Since this is all in your control, and get’s nothing from the framework.

What do you think?
Tom

**Comment by tomweber-sas** [comment]
Hey Debbie, any updates on this?

Thanks,
Tom

**Comment by tomweber-sas** [comment]
Hey Debbie, I hope what I suggested worked, or maybe the newer version of that package you are using provided a way to access an object from the functions. I'm going to close this, as there's nothing that can be done on my end, other than the suggestion I had about how to implement this in your framework. But if there's anything else you need, feel free to reopen it or open another for anything else too. 
Thanks!
Tom

**Comment by Debbby57** [comment]
Hello @tomweber-sas,
I am really sorry for the late reply, I was on other urgent subjects.

We discussed about your solution, but we can't use it because the nodes don't have access to credentials. If we do that, we have to send credentials in the nodes and we are not confortable with this.
We decided to upgrade our project to the 0.16.5 version of our framework. It doesn't resolve our problem, but with this version it will be easier to force a "global" SAS Session which can be used in our pipelines, and we think it will be more maintanable if we don't affect nodes and pipelines. I'm going to try this in the coming weeks.

I really thank you for taking the time for my problem. If you want to have updates about this, let me know and I will add some comments in this issue ;)
